src/input/json_to_storage.cc

```diff
@@ -22,46 +22,184 @@ using namespace internal;
+                THROW( ExcReferenceNotFound() << EI_RefAddress(this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
+                        << EI_Specification("there should be Record") );
+            if ( !ref_path->down(tmp_str) )
+                THROW( ExcReferenceNotFound() << EI_RefAddress(this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
+                        << EI_Specification("key '"+tmp_str+"' not found") );
+        }
+        pos = new_pos+1;
+    }
+    if (relative_ref) {
+    	xprintf(Msg, "Referencing '%s' to '%s'.\n", this->str().c_str(), ref_path->str().c_str());
+    }
+    return ref_path;
+}
+
+
+
+/********************************************
+ * Implementation of  internal::PathJSON
+ */
+
+
+PathJSON::PathJSON(istream &in)
+: PathJSON()
+{
+    io::filtering_istream filter_in;
+
+    filter_in.push(uncommenting_filter());
+    filter_in.push(in);
+
+    Node root_node;
+
+    try {
+        json_spirit::read_or_throw( filter_in, root_node);
+    } catch (json_spirit::Error_position &e ) {
+        THROW( JSONToStorage::ExcNotJSONFormat() << JSONToStorage::EI_JSONLine(e.line_) << JSONToStorage::EI_JSONColumn(e.column_)
+        	<< JSONToStorage::EI_JSONReason(e.reason_));
+    }
+
+    nodes_.push_back( new Node(root_node) );
 }
 
-const JSONPath::Node * JSONPath::down(unsigned int index)
+
+PathJSON::PathJSON()
+: PathBase()
+{
+    /* from json_spirit_value.hh:
+     * enum Value_type{ obj_type, array_type, str_type, bool_type, int_type, real_type, null_type };
+     */
+    json_type_names.push_back("JSON object");
+    json_type_names.push_back("JSON array");
+    json_type_names.push_back("JSON string");
+    json_type_names.push_back("JSON bool");
+    json_type_names.push_back("JSON int");
+    json_type_names.push_back("JSON real");
+    json_type_names.push_back("JSON null");
+}
+
+
+bool PathJSON::down(unsigned int index)
 {
     const Node * head_node = nodes_.back();
     const json_spirit::mArray &array = head_node->get_array(); // the type should be checked in make_storage
 
-    if (  index >= array.size()) return NULL;
+    if (  index >= array.size()) return false;
     path_.push_back( make_pair( index, string("") ) );
     nodes_.push_back( &( array[index] ) );
 
-    return nodes_.back();
+    return (nodes_.back() != NULL);
 }
 
 
-const JSONPath::Node * JSONPath::down(const string& key)
+bool PathJSON::down(const string& key)
 {
     const Node * head_node = nodes_.back();
     const json_spirit::mObject &obj = head_node->get_obj(); // the type should be checked in make_storage
 
     json_spirit::mObject::const_iterator it = obj.find(key);
     if (it == obj.end()) {
-        return NULL;
+        return false;
     } else {
         path_.push_back( make_pair( (int)(-1), key) );
         nodes_.push_back( &( it->second ) );
     }
-    return nodes_.back();
+    return (nodes_.back() != NULL);
 }
 
 
-void JSONPath::up()
+void PathJSON::up()
 {
     if (path_.size() > 1) {
```

Comment:
We can delete node only in PathYAML::up() where copy of node is stored to nodes vector. In PathJSON is directly stored nodes from json_spirit tree and we can't delete them (it throws segfault).